lib/DirManager/dirman_vitafs.cpp

```diff
@@ -0,0 +1,493 @@
+
+The above copyright notice and this permission notice shall be included
+in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+*/
+
+
+#ifdef VITA
+#include <sys/dirent.h>
+
+#include <psp2/io/devctl.h>
+#include <psp2/io/dirent.h>
+#include <psp2/io/fcntl.h>
+#include <psp2/io/stat.h>
+#include <sys/fcntl.h>
+#include <dirent.h>
+#endif
+
+#include <unistd.h>
+#include <memory.h>
+
+
+
+#include "dirman.h"
+#include "dirman_private.h"
+#include <mutex>
+#include <limits.h>
+
+#include <Logger/logger.h>
+
+#ifndef PATH_MAX
+#define PATH_MAX 1024
+#endif
+
+#ifndef MAX_PATH_LENGTH
+#define MAX_PATH_LENGTH PATH_MAX
+#endif
+
+
+
+static std::mutex g_dirManMutex;
+
+#define PUT_THREAD_GUARD() \
+    std::lock_guard<std::mutex> guard(g_dirManMutex);\
+    (void)guard;
+
+void DirMan::DirMan_private::setPath(const std::string &dirPath)
+{
+    PUT_THREAD_GUARD()
+#ifdef VITA
+    if(dirPath.empty())
+    {
+        pLogWarning("[dirman_vitafs] WARNING: ::setPath called with dirPath of %s", dirPath.c_str());
+        return;
+    }
+    m_dirPath = dirPath;
+    delEnd(m_dirPath, '/');
+#endif
+}
+
+static inline int hasEndSlash(char* string)
+{
+    if(string == NULL) return -1;
+    return string[strlen(string) - 1] == '/';
+}
+
+bool DirMan::DirMan_private::getListOfFiles(std::vector<std::string> &list, const std::vector<std::string> &suffix_filters)
+{
+    PUT_THREAD_GUARD()
+    list.clear();
+
+    // open directory fd
+    SceUID dfd = sceIoDopen(m_dirPath.c_str());
+    if(dfd >= 0)
+    {
+        int res = 0;
+        do
+        {
+            SceIoDirent dirEntry;
+            memset(&dirEntry, 0, sizeof(SceIoDirent));
+
+            res = sceIoDread(dfd, &dirEntry);
+            if(res > 0)
+            {
+                char *new_path = (char*)malloc(strlen(m_dirPath.c_str()) + strlen(dirEntry.d_name) + 2);
+                snprintf(new_path, MAX_PATH_LENGTH, "%s%s%s", m_dirPath.c_str(), hasEndSlash((char*)m_dirPath.c_str()) ? "" : "/", dirEntry.d_name);
+                // pLogDebug("Discovered new FILE path `%s`", new_path);
+
+                // matching non-directories
+                if (!SCE_S_ISDIR(dirEntry.d_stat.st_mode))
+                {
+                    if(matchSuffixFilters(dirEntry.d_name , suffix_filters))
+                            list.push_back(dirEntry.d_name);
+                }
+
+                free(new_path);
+            }
+            /* code */
+        } while (res > 0);
+        
+        sceIoDclose(dfd);
+    }
+
+
+
+    // POSIX
```

Comment:
As you copied a file for Vita exclusive support, you can freely get rid of all commented code as it still being at "posix" file if need some for hints :)

Reply:
Whoops. Yeah that commented out code was useful when I was originally writing it on my laptop.